pytorch_lightning/loops/batch_loop.py

```diff
@@ -0,0 +1,548 @@
+from collections import OrderedDict
+from contextlib import contextmanager
+from copy import copy
+from functools import partial, update_wrapper
+from typing import Any, Callable, List, Mapping, Optional, Tuple
+
+import numpy as np
+import torch
+from torch.optim import Optimizer
+
+from pytorch_lightning.core.optimizer import LightningOptimizer
+from pytorch_lightning.loops.base import Loop
+from pytorch_lightning.plugins import ParallelPlugin
+from pytorch_lightning.trainer.supporters import TensorRunningAccum
+from pytorch_lightning.utilities import AMPType, AttributeDict, DeviceType, grad_norm
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.finite_checks import detect_nan_parameters
+from pytorch_lightning.utilities.imports import _TPU_AVAILABLE
+from pytorch_lightning.utilities.signature_utils import is_param_in_hook_signature
+from pytorch_lightning.utilities.warnings import WarningCache
+
+
+# TODO: typing
+class BatchLoop(Loop):
+    """ Runs over a single batch of data. """
+
+    def __init__(self):
+        super().__init__()
+        self.accumulated_loss = None
+        self.running_loss = TensorRunningAccum(window_length=20)
+        self.split_idx = None
+        self.warning_cache = WarningCache()
+
+        self._hiddens = None
+        self._optimizer_freq_cumsum = None
+        self._skip_backward = False
+
+    @property
+    def skip_backward(self) -> bool:
+        """ Determines whether the loop will skip backward during automatic optimization. """
+        return self._skip_backward
+
+    @skip_backward.setter
+    def skip_backward(self, value: bool):
+        """ Determines whether the loop will skip backward during automatic optimization. """
+        self._skip_backward = value
```

Comment:
this is only relevant during training. does this warrant a separate TrainBatchLoop vs regular batch loop?

Reply:
Maybe yes. I think further refinement would be good. Right now I don't yet see a general batch loop for evaluation, because there is nothing to loop over, unlike in training where we have optimizers and tbptt